File: src/samples/blackscreen/blackscreen.cpp
// mge - Modern Game Engine
// Copyright (c) 2017-2023 by Alexander Schroeder
// All rights reserved.
#include "mge/application/application.hpp"
#include "mge/core/trace.hpp"
#include "mge/graphics/render_context.hpp"
#include "mge/graphics/render_system.hpp"
#include "mge/graphics/rgba_color.hpp"
#include "mge/graphics/window.hpp"

namespace mge {
    MGE_DEFINE_TRACE(BLACKSCREEN);
}
namespace mge {
    class blackscreen : public application
    {
    public:
        blackscreen() = default;

        void setup() override
        {
            MGE_DEBUG_TRACE(BLACKSCREEN, "Setup blackscreen");
            m_render_system = render_system::create();
            m_window = m_render_system->create_window();
            m_window->set_close_listener([&] {
                MGE_DEBUG_TRACE(BLACKSCREEN, "Close listener invoked");
                set_quit();
            });
            m_window->set_key_action_handler(
                [&](mge::key k, mge::key_action a, mge::modifier m) -> bool {
                    if (a == mge::key_action::PRESS && k == mge::key::ESCAPE) {
                        set_quit();
                        return true;
                    }
                    return false;
                });

            add_redraw_listener([&](uint64_t cycle, double delta) {
                auto& pass = m_window->render_context().pass(0);
                pass.default_viewport();
                pass.clear_color(mge::rgba_color(0.0f, 0.0f, 0.0f, 1.0f));
                pass.touch();

                m_window->render_context().frame();
            });

            m_window->show();
        }

        void teardown() override
        {
            m_window.reset();
            m_render_system.reset();
        }

    private:
        render_system_ref m_render_system;
        window_ref        m_window;
    };

    MGE_REGISTER_IMPLEMENTATION(blackscreen, mge::application, blackscreen);
} // namespace mge

MGE_MAINFUNCTION

File: src/samples/blackscreen_lua/blackscreen.lua
-- blackscreen sample in Lua
-- Recreates the C++ blackscreen sample using the Lua scripting bindings.

local BlackScreen = class(mge.application)
BlackScreen.__name__ = "blackscreen_lua"

local render_system
local window

function BlackScreen:setup()
    render_system = mge.render_system.create()
    window = render_system:create_window()

    window:set_close_listener(function()
        self:set_quit()
    end)

    window:add_key_action_handler(function(k, action, modifier)
        if action == mge.key_action.PRESS and k == mge.key.ESCAPE then
            self:set_quit()
            return true
        end
        return false
    end)

    self:add_redraw_listener(function(cycle, delta)
        local ctx = window:render_context()
        local p = ctx:pass(0)
        p:default_viewport()
        p:clear_color(mge.rgba_color(0.0, 0.0, 0.0, 1.0))
        p:touch()
        ctx:frame()
    end)

    window:show()
end

function BlackScreen:teardown()
    window = nil
    render_system = nil
    collectgarbage("collect")
end

mge.register_component(mge.application, BlackScreen)
local rc = mge.application.main("blackscreen_lua")
return rc == 0
